### DISCUSSION TITLE
Trouble with loading SMILES to QED module

### DISCUSSION THREAD

**Discussion by ShirosakiK** [rdkit/rdkit#7158] [Q&A] [answered]
Good day!

I am a Biochemistry undergraduate, and for my thesis I am working on designing bioinorganic complexes in silico and looking at their possible applications in medicine.

In any case, the complexes I'm working with contain an aromatic ligand sandwiched to the metal center and as you may know drawing such a bond  is still a limitation in many drawing softwares. As a compromise, I directly bonded the carbons to the metal center. Subsequently, I generated the SMILES and then tried calculating its QED in RDKit. However, I get the following error even though the valence of the carbons seem to check out (please correct me if I am wrong on this). I have also tried standardizing the SMILES input but I also get the error shown.

____________________

>>> from qed import qed
>>> from rdkit import Chem
>>> m=Chem.MolFromSmiles('CC(C)C12=C3C4=C5(C)C6=C1[Ru]143625(Cl)O=C(CC(=O1)/C=C/c1ccc(OC)c(OC)c1)/C=C/c1ccc(O)c(OC)c1')
[07:29:03] Explicit valence for atom # 3 C, 5, is greater than permitted
>>> qed.default(m)
Error: in function properties(mol): mol argument is 'None'

____________________

I have also tried setting sanitize=False, and while I am now able to use the QED module, I get the following error.

____________________

>>> from qed import qed
>>> from rdkit import Chem
>>> m=Chem.MolFromSmiles('CC(C)C12=C3C4=C5(C)C6=C1[Ru]143625(Cl)O=C(CC(=O1)/C=C/c1ccc(OC)c(OC)c1)/C=C/c1ccc(O)c(OC)c1', sanitize=False)
>>> qed.default(m)
[07:32:01]

****
Pre-condition Violation
getNumImplicitHs() called without preceding call to calcImplicitValence()
Violation occurred on line 297 in file /home/conda/feedstock_root/build_artifacts/rdkit_1660289032136/work/Code/GraphMol/Atom.cpp
Failed Expression: d_implicitValence > -1
****

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/anaconda3/envs/rdkit/lib/python3.10/site-packages/qed/qed.py", line 269, in default
    return weights_mean(mol, gerebtzoff)
  File "/home/user/anaconda3/envs/rdkit/lib/python3.10/site-packages/qed/qed.py", line 251, in weights_mean
    props = properties(mol)
  File "/home/user/anaconda3/envs/rdkit/lib/python3.10/site-packages/qed/qed.py", line 207, in properties
    x[0] = Descriptors.MolWt(mol)                                                                                      # MW
  File "/home/user/anaconda3/envs/rdkit/lib/python3.10/site-packages/rdkit/Chem/Descriptors.py", line 69, in <lambda>
    MolWt = lambda *x, **y: _rdMolDescriptors._CalcMolWt(*x, **y)
RuntimeError: Pre-condition Violation
        getNumImplicitHs() called without preceding call to calcImplicitValence()
        Violation occurred on line 297 in file Code/GraphMol/Atom.cpp
        Failed Expression: d_implicitValence > -1
        RDKIT: 2022.03.5
        BOOST: 1_74

____________________

Any help on how to resolve this error would be much appreciated. Thanks!

**Comment by ptosco** [comment]
@ShirosakiK Please also paste your code as text, or people willing to help will have to retype the code from your image.

**Reply by ShirosakiK** [reply]
I've updated the post to reflect the code and the error message. Thanks.

**Comment by ptosco** [comment, marked as answer]
@ShirosakiK
An option to preserve the integrity of the complex while avoiding to exceed the valences of carbon, nitrogen and oxygen atom involved in such complexes is to convert those single bonds into zero-order bonds:

```
from rdkit import Chem

m = Chem.MolFromSmiles('CC(C)C12=C3C4=C5(C)C6=C1[Ru]143625(Cl)O=C(CC(=O1)/C=C/c1ccc(OC)c(OC)c1)/C=C/c1ccc(O)c(OC)c1', sanitize=False)

for b in m.GetBonds():
    ru = None
    if b.GetBeginAtom().GetAtomicNum() == 44:
        ru = b.GetBeginAtom()
    elif b.GetEndAtom().GetAtomicNum() == 44:
        ru = b.GetEndAtom()
    if ru is None:
        continue
    other_atom = b.GetOtherAtom(ru)
    if other_atom.GetAtomicNum() in (6, 7, 8):
        b.SetBondType(Chem.BondType.ZERO)

Chem.SanitizeMol(m)
rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE

Chem.MolToSmiles(m)
'COc1cc(/C=C/C2CC(/C=C/c3ccc(OC)c(OC)c3)=O~[Ru]34567(Cl)(~O=2)~c2c~3c~4(C(C)C)c~5c~6c~72C)ccc1O'
```
Alternatively, if you wish to preserve the dative nature of the coordination bond, you may delete the existing single bonds and re-add a dative bond pointing from the non-metal to the metal:
```
from rdkit import Chem

m = Chem.MolFromSmiles('CC(C)C12=C3C4=C5(C)C6=C1[Ru]143625(Cl)O=C(CC(=O1)/C=C/c1ccc(OC)c(OC)c1)/C=C/c1ccc(O)c(OC)c1', sanitize=False)

bonds_to_add = []
with Chem.RWMol(m) as m:
    for b in m.GetBonds():
        ru = None
        if b.GetBeginAtom().GetAtomicNum() == 44:
            ru = b.GetBeginAtom()
        elif b.GetEndAtom().GetAtomicNum() == 44:
            ru = b.GetEndAtom()
        if ru is None:
            continue
        other_atom = b.GetOtherAtom(ru)
        if other_atom.GetAtomicNum() in (6, 7, 8):
            m.RemoveBond(ru.GetIdx(), other_atom.GetIdx())
            bonds_to_add.append((other_atom.GetIdx(), ru.GetIdx(), Chem.BondType.DATIVE))

with Chem.RWMol(m) as m:
    for b in bonds_to_add:
        m.AddBond(*b)

Chem.SanitizeMol(m)
rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE

Chem.MolToSmiles(m)
'COc1cc(/C=C/C2CC(/C=C/c3ccc(OC)c(OC)c3)=O->[Ru]34567(Cl)(<-O=2)<-c2c->3c->4(C(C)C)c->5c->6c->72C)ccc1O'
```

**Reply by ShirosakiK** [reply]
Thank you! I am now able to make the QED module work.
